Boost Math bernoulli code fails compile on platforms with no atomic ints. This patch
fixes that at the expense of multithread safety


diff --git a/boost/math/special_functions/detail/bernoulli_details.hpp b/boost/math/special_functions/detail/bernoulli_details.hpp
index cf35545264..8519b7c89c 100644
--- a/boost/math/special_functions/detail/bernoulli_details.hpp
+++ b/boost/math/special_functions/detail/bernoulli_details.hpp
@@ -360,7 +360,7 @@ class bernoulli_numbers_cache
          return out;
       }
 
-      #ifndef BOOST_HAS_THREADS
+      #if !defined(BOOST_HAS_THREADS) || defined(BOOST_MATH_BERNOULLI_UNTHREADED)
       //
       // Single threaded code, very simple:
       //
@@ -382,6 +382,8 @@ class bernoulli_numbers_cache
          *out = (i >= m_overflow_limit) ? policies::raise_overflow_error<T>("boost::math::bernoulli_b2n<%1%>(std::size_t)", 0, T(i), pol) : bn[i];
          ++out;
       }
+      #elif defined(BOOST_MATH_NO_ATOMIC_INT)
+      static_assert(sizeof(T) == 1, "Unsupported configuration: your platform appears to have no atomic integers.  If you are happy with thread-unsafe code, then you may define BOOST_MATH_BERNOULLI_UNTHREADED to suppress this error.");
       #else
       //
       // Double-checked locking pattern, lets us access cached already cached values
@@ -464,7 +466,7 @@ class bernoulli_numbers_cache
          return out;
       }
 
-      #ifndef BOOST_HAS_THREADS
+      #if !defined(BOOST_HAS_THREADS) || defined(BOOST_MATH_BERNOULLI_UNTHREADED)
       //
       // Single threaded code, very simple:
       //
@@ -494,6 +496,8 @@ class bernoulli_numbers_cache
          }
          ++out;
       }
+      #elif defined(BOOST_MATH_NO_ATOMIC_INT)
+      static_assert(sizeof(T) == 1, "Unsupported configuration: your platform appears to have no atomic integers.  If you are happy with thread-unsafe code, then you may define BOOST_MATH_BERNOULLI_UNTHREADED to suppress this error.");
       #else
       //
       // Double-checked locking pattern, lets us access cached already cached values
@@ -555,7 +559,7 @@ class bernoulli_numbers_cache
    // The value at which we know overflow has already occurred for the Bn:
    std::size_t m_overflow_limit;
 
-   #ifdef BOOST_HAS_THREADS
+   #if defined(BOOST_HAS_THREADS) && !defined(BOOST_MATH_NO_ATOMIC_INT)
    std::mutex m_mutex;
    atomic_counter_type m_counter, m_current_precision;
    #else
diff --git a/boost/math/tools/atomic.hpp b/boost/math/tools/atomic.hpp
index cc76ed269f..e3cbf5db89 100644
--- a/boost/math/tools/atomic.hpp
+++ b/boost/math/tools/atomic.hpp
@@ -16,27 +16,27 @@
 namespace boost {
    namespace math {
       namespace detail {
-#if ATOMIC_INT_LOCK_FREE == 2
+#if (ATOMIC_INT_LOCK_FREE == 2) && !defined(BOOST_MATH_NO_ATOMIC_INT)
          typedef std::atomic<int> atomic_counter_type;
          typedef std::atomic<unsigned> atomic_unsigned_type;
          typedef int atomic_integer_type;
          typedef unsigned atomic_unsigned_integer_type;
-#elif ATOMIC_SHORT_LOCK_FREE == 2
+#elif (ATOMIC_SHORT_LOCK_FREE == 2) && !defined(BOOST_MATH_NO_ATOMIC_INT)
          typedef std::atomic<short> atomic_counter_type;
          typedef std::atomic<unsigned short> atomic_unsigned_type;
          typedef short atomic_integer_type;
          typedef unsigned short atomic_unsigned_type;
-#elif ATOMIC_LONG_LOCK_FREE == 2
+#elif (ATOMIC_LONG_LOCK_FREE == 2) && !defined(BOOST_MATH_NO_ATOMIC_INT)
          typedef std::atomic<long> atomic_unsigned_integer_type;
          typedef std::atomic<unsigned long> atomic_unsigned_type;
          typedef unsigned long atomic_unsigned_type;
          typedef long atomic_integer_type;
-#elif ATOMIC_LLONG_LOCK_FREE == 2
+#elif (ATOMIC_LLONG_LOCK_FREE == 2) && !defined(BOOST_MATH_NO_ATOMIC_INT)
          typedef std::atomic<long long> atomic_unsigned_integer_type;
          typedef std::atomic<unsigned long long> atomic_unsigned_type;
          typedef long long atomic_integer_type;
          typedef unsigned long long atomic_unsigned_integer_type;
-#else
+#elif !defined(BOOST_MATH_NO_ATOMIC_INT)
 #  define BOOST_MATH_NO_ATOMIC_INT
 #endif
       } // Namespace detail
diff --git a/libs/math/test/Jamfile.v2 b/libs/math/test/Jamfile.v2
index 52fb87f5e5..7386b4ff84 100644
--- a/libs/math/test/Jamfile.v2
+++ b/libs/math/test/Jamfile.v2
@@ -1145,6 +1145,9 @@ test-suite misc :
 
 #   [ run __temporary_test.cpp test_instances//test_instances : : : <test-info>always_show_run_output <pch>off ]
    [ compile test_no_long_double_policy.cpp ]
+   [ compile bernoulli_no_atomic_d.cpp ]
+   [ compile bernoulli_no_atomic_mp.cpp ]
+   [ compile-fail bernoulli_no_atomic_fail.cpp ]
 ;
 
 test-suite interpolators :
diff --git a/libs/math/test/bernoulli_no_atomic_d.cpp b/libs/math/test/bernoulli_no_atomic_d.cpp
new file mode 100644
index 0000000000..a9c37d60cb
--- /dev/null
+++ b/libs/math/test/bernoulli_no_atomic_d.cpp
@@ -0,0 +1,14 @@
+//  (C) Copyright John Maddock 2021.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#define BOOST_MATH_NO_ATOMIC_INT
+
+#include <boost/math/special_functions/bernoulli.hpp>
+#include "compile_test/test_compile_result.hpp"
+
+void compile_and_link_test()
+{
+   check_result<double>(boost::math::bernoulli_b2n<double>(4));
+}
diff --git a/libs/math/test/bernoulli_no_atomic_fail.cpp b/libs/math/test/bernoulli_no_atomic_fail.cpp
new file mode 100644
index 0000000000..d1f4faa387
--- /dev/null
+++ b/libs/math/test/bernoulli_no_atomic_fail.cpp
@@ -0,0 +1,15 @@
+//  (C) Copyright John Maddock 2021.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#define BOOST_MATH_NO_ATOMIC_INT
+
+#include <boost/math/special_functions/bernoulli.hpp>
+#include <boost/multiprecision/cpp_bin_float.hpp>
+#include "compile_test/test_compile_result.hpp"
+
+void compile_and_link_test()
+{
+   check_result<boost::multiprecision::cpp_bin_float_50>(boost::math::bernoulli_b2n<boost::multiprecision::cpp_bin_float_50>(4));
+}
diff --git a/libs/math/test/bernoulli_no_atomic_mp.cpp b/libs/math/test/bernoulli_no_atomic_mp.cpp
new file mode 100644
index 0000000000..53cb5ffed0
--- /dev/null
+++ b/libs/math/test/bernoulli_no_atomic_mp.cpp
@@ -0,0 +1,16 @@
+//  (C) Copyright John Maddock 2021.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#define BOOST_MATH_NO_ATOMIC_INT
+#define BOOST_MATH_BERNOULLI_UNTHREADED
+
+#include <boost/math/special_functions/bernoulli.hpp>
+#include <boost/multiprecision/cpp_bin_float.hpp>
+#include "compile_test/test_compile_result.hpp"
+
+void compile_and_link_test()
+{
+   check_result<boost::multiprecision::cpp_bin_float_50>(boost::math::bernoulli_b2n<boost::multiprecision::cpp_bin_float_50>(4));
+}
